marlowe-dashboard-client/src/Contract/State.purs

```diff
@@ -60,12 +58,184 @@ import Web.HTML.HTMLElement as HTMLElement
 
 -- see note [dummyState] in MainFrame.State
 dummyState :: State
-dummyState = mkInitialState emptyContractInstanceId zero emptyContractMetadata mempty Nothing Close
+dummyState =
+  { tab: Tasks
+  , executionState: initExecution zero contract
+  , previousSteps: mempty
+  , marloweParams: emptyMarloweParams
+  , contractInstanceId: emptyContractInstanceId
+  , selectedStep: 0
+  , metadata: emptyMetaData
+  , participants: mempty
+  , mActiveUserParty: Nothing
+  , namedActions: mempty
+  }
+  where
+  contract = Close
+
+  emptyContractInstanceId = ContractInstanceId UUID.emptyUUID
+
+  emptyMarloweParams = { rolePayoutValidatorHash: mempty, rolesCurrency: CurrencySymbol { unCurrencySymbol: "" } }
+
+  emptyMarloweData = { marloweContract: contract, marloweState: emptyMarloweState }
+
+  emptyMarloweState = Semantic.State { accounts: mempty, choices: mempty, boundValues: mempty, minSlot: zero }
+
+  emptyMetaData =
+    { contractType: Escrow
+    , contractName: mempty
+    , contractDescription: mempty
+    , roleDescriptions: mempty
+    , slotParameterDescriptions: mempty
+    , valueParameterDescriptions: mempty
+    , choiceDescriptions: mempty
+    }
+
+mkInitialState :: Slot -> ContractInstanceId -> History -> Maybe State
+mkInitialState currentSlot contractInstanceId history =
+  let
+    marloweParams = get1 $ unwrap history
+
+    marloweData = get2 $ unwrap history
+
+    transactionInputs = get3 $ unwrap history
+
+    contract = marloweData.marloweContract
+
+    mTemplate = findTemplate contract
+
+    -- We can't use the currentSlot to create the initial execution state, since the contract might
+    -- have been created several slots ago. Hopefully this doesn't matter (the argument is only used
+    -- to set the minSlot in the contract's initial state).
+    initialExecutionState = initExecution zero contract
+  in
+    flip map mTemplate \template ->
+      let
+        initialState =
+          { tab: Tasks
+          , executionState: initialExecutionState
+          , previousSteps: mempty
+          , marloweParams
+          , contractInstanceId
+          , selectedStep: 0
+          , metadata: template.metaData
+          , participants: mempty --- FIXME: I don't think there is any way we can do this
+          , mActiveUserParty: Nothing -- FIXME: this should be a function of the walletDetails
+          , namedActions: mempty
+          }
+
+        updateExecutionState = over _executionState (applyTransactionInputs transactionInputs)
+      in
+        initialState
+          # updateExecutionState
+          # regenerateStepCards currentSlot
+          # selectLastStep
+
+updateState :: Slot -> History -> State -> State
+updateState currentSlot history state =
+  let
+    allTransactionInputs = get3 $ unwrap history
+
+    previousTransactionInputs = map (\previousState -> previousState.txInput) $ view (_executionState <<< _previousState) state
```

Comment:
I'd have to double check, but as long as we are using lenses and if the data is what I think it is, this could be replaced with 

```purescript
previousTransactionInputs = toArrayOf (_executionState <<< _previousTransactions) state
```

Note that `_previousTransactions` is in Marlowe.Execution

Reply:
:+1: